File: smart-vial-backend/models/EmaResponse.js
const mongoose = require("mongoose");
const { GROUND_TRUTH } = require("../utils/groundTruthMetrics");

// ============================================
// GTM-521 — EMA self-report (ground-truth label source)
// ============================================
// An EmaResponse is one ecological-momentary-assessment self-report: the patient
// answering "Did you just take your dose?" (yes / no / unsure) close in time to a
// detected dose interaction. This is the SUPERVISED LABEL we validate the
// device's dose DETECTION against — crossing the self-report (ground truth) with
// the DoseEvent (prediction) yields dose-detection sensitivity/specificity
// (utils/groundTruthMetrics.js).
//
// PHI-free: subjects are referenced by id only (user_id / device_id, both
// server-derived from the session-owned device — never client-asserted). There is
// no medication, name, free text, or condition field anywhere here; the only
// payload is the constrained self-report enum + sampling metadata.
//
// Idempotent: a client-supplied `idempotency_key` (unique, sparse) lets a retried
// submission de-dupe to exactly one row, matching the dose-event ingest path.

const EmaResponseSchema = new mongoose.Schema(
  {
    // ----- Subject reference (ids only — never PHI) -----
    // The owning user (session id, server-derived). Lets a response be scoped /
    // purged per subject without ever storing a name.
    user_id: {
      type: String,
      required: true,
      index: true,
    },
    // The device the self-report concerns. Bound from the session-owned device by
    // the controller; the client cannot assert it.
    device_id: {
      type: String,
      required: true,
      index: true,
    },
    // Optional link to the specific detected DoseEvent this prompt followed. When
    // present, the metrics engine can pair this response to that exact detection;
    // when absent, it is matched by time window. Opaque id, not PHI.
    dose_event_id: {
      type: String,
      default: null,
      index: true,
    },

    // ----- The self-report (the ground-truth label) -----
    // Constrained enum — yes/no/unsure. "unsure" is a real answer but is NOT
    // usable ground truth (excluded from the confusion matrix by the metrics
    // engine), so we record it rather than forcing a guess.
    self_reported_taken: {
      type: String,
      required: true,
      enum: [GROUND_TRUTH.TAKEN, GROUND_TRUTH.NOT_TAKEN, GROUND_TRUTH.UNSURE],
    },

    // ----- Sampling metadata (PHI-free) -----
    // When the EMA prompt was shown and when it was answered. `prompted_at` is
    // set from sub-sampling; `responded_at` from the submission.
    prompted_at: {
      type: Date,
      default: null,
    },
    responded_at: {
      type: Date,
      default: Date.now,
      index: true,
    },
    // Opaque tag describing WHY this dose event was sampled for a prompt (e.g.
    // "subsample"), plus the sampling rate in effect. Numbers/codes only.
    sampling: {
      type: new mongoose.Schema(
        {
          reason: { type: String, default: null },
          rate: { type: Number, default: null },
        },
        { _id: false }
      ),
      default: () => ({}),
    },

    // ----- Idempotency -----
    idempotency_key: {
      type: String,
      unique: true,
      sparse: true,
      index: true,
    },
  },
  { timestamps: { createdAt: "createdAt", updatedAt: false } }
);

// Query a subject's recent EMA responses for the metrics window (ids only).
EmaResponseSchema.index({ user_id: 1, responded_at: -1 });
EmaResponseSchema.index({ device_id: 1, responded_at: -1 });

module.exports = mongoose.model("EmaResponse", EmaResponseSchema);

File: farda-app/lib/application/device/model/ema_response.dart
/// GTM-521 — EMA (ecological momentary assessment) self-report.
///
/// After a dose-event sync, on a server-decided SUB-SAMPLE, the app asks "Did you
/// just take your dose?" (yes / no / unsure). The answer is the SUPERVISED LABEL
/// the Vial backend validates device dose-detection against
/// (smart-vial-backend `utils/groundTruthMetrics.js`), so we keep it PHI-free:
/// one constrained enum value + the opaque `dose_event_id` it followed.
///
/// This model is pure (no widgets / network) so the answer mapping and the wire
/// body shape are unit-testable under `flutter test` in CI.
enum EmaAnswer {
  /// "Yes, I took it" ⇒ ground-truth label `taken`.
  taken,

  /// "No, I didn't" ⇒ ground-truth label `not_taken`.
  notTaken,

  /// "Not sure" ⇒ a real answer, but NOT usable ground truth (the backend counts
  /// it separately, never guessing a label).
  unsure,
}

extension EmaAnswerWire on EmaAnswer {
  /// The exact string the Vial `recordEmaResponse` endpoint expects for
  /// `self_reported_taken` (matches `GROUND_TRUTH` in groundTruthMetrics.js).
  String get wireValue {
    switch (this) {
      case EmaAnswer.taken:
        return 'taken';
      case EmaAnswer.notTaken:
        return 'not_taken';
      case EmaAnswer.unsure:
        return 'unsure';
    }
  }
}

/// One EMA self-report ready to POST to the Vial backend.
class EmaResponse {
  /// The vial this self-report concerns (path-bound on the backend too; sent so
  /// the body is self-describing).
  final String deviceId;

  /// The patient's answer to the prompt.
  final EmaAnswer answer;

  /// The detected dose event this prompt followed, if known (server-issued in the
  /// dose-event ingest response's `ema_prompt.dose_event_id`). Lets the backend
  /// pair this label to that exact detection.
  final String? doseEventId;

  const EmaResponse({
    required this.deviceId,
    required this.answer,
    this.doseEventId,
  });

  /// Stable idempotency key so a retried submission de-dupes to one row on the
  /// backend. Derived from the device + the dose it answers (NOT wall-clock), so
  /// re-sending the same answer for the same dose can't double-count.
  String get idempotencyKey =>
      'ema_${deviceId}_${doseEventId ?? 'na'}_${answer.wireValue}';

  /// The exact JSON body the Vial `recordEmaResponse` endpoint expects. Extracted
  /// so the payload shape is unit-testable without any network.
  Map<String, dynamic> toBody() {
    return {
      'self_reported_taken': answer.wireValue,
      if (doseEventId != null) 'dose_event_id': doseEventId,
      'idempotency_key': idempotencyKey,
    };
  }
}
